### ISSUE TITLE
Invalid response from metadata service: incorrect Metadata-Flavor header

### ISSUE THREAD

**Issue by arianlopezc** [googleapis/nodejs-logging#561]
After reviewing as much as I could online to see how I could solve the issue, I wasn't able to come to a solution, this is an issue apparently with how the JSON file holding the credentials to GCloud.

"Invalid response from metadata service: incorrect Metadata-Flavor header"

That is normally the error that keeps throwing the client when I invoke the alert function of the client.

#### Environment details

  - OS: macOS Mojave
  - Node.js version: v10.15.2
  - npm version: 6.10.0
  - `@google-cloud/logging` version: 5.2.1

#### Steps to reproduce

  1. Generate a quick Angular project using the CLI
  2. Make sure to have a JSON file with service account credentials properly downloaded.
  3. Generate a quick service that looks somehow like it follows, inject it, and call the function that performs the logging.

`import { Logging } from '@google-cloud/logging';
@Injectable()
export class GCloudLogService {
         constructor(
		private readonly logging: Logging,
	) {}

        async saveLog(logPayload: IServerExceptionLog) {
	try {
		const metadata = {
			resource: { type: 'global' },
		  };
		const entry = this.logging.entry(metadata, logPayload);
		const log = this.logging.log(logName);
		await log.alert(entry);
		return;
	} catch (err) {
		console.log(`ERR ${err.message}`);
	}
  }
}`

No matter how many versions of the code I try I keep getting the error:

"Invalid response from metadata service: incorrect Metadata-Flavor header."

I tried setting the project manually,

await this.logging.setProjectId({
			projectId,
		});

but still nothing.

I did make sure that the environment variable GOOGLE_APPLICATION_CREDENTIALS was set up and it is.

Please help, I have expended too many hours trying to figure it out but still nothing.

**Comment by arianlopezc** [comment]
Hi @bcoe, do you by any chance have any recommendation on what could I do?

**Comment by bcoe** [comment]
@arianlopezc :wave: I don't know a solution off the top of my head, and will need to dig into the problem a bit.

Will work on getting back to you as soon as possible.

**Comment by arianlopezc** [comment]
@bcoe 👋is there another way I could push the logs? Maybe could use the API instead of the library... do you believe that could be a proper "secondary option" in the mean time?

**Comment by bcoe** [comment]
@arianlopezc as a starting point, have you tried simply running the code in `samples/quickstart.js`?

It works like a charm for me without any exceptions:

```js
  // Imports the Google Cloud client library
  const {Logging} = require('@google-cloud/logging');

  // Creates a client
  const logging = new Logging({projectId});

  // Selects the log to write to
  const log = logging.log(logName);

  // The data to write to the log
  const text = 'Hello, world!';

  // The metadata associated with the entry
  const metadata = {
    resource: {type: 'global'},
  };

  // Prepares a log entry
  const entry = log.entry(metadata, text);

  // Writes the log entry
  await log.write(entry);
  console.log(`Logged: ${text}`);
}
```

**Comment by arianlopezc** [comment]
@bcoe I tried it again just in case, step y step, but still nothing. The issue with the example is that it is for regular .js, while I'm trying this using Typescript in Angular and also NestJS.
I started putting some logs for the console and noticed that if I used the line from the example in my function:
`const {Logging} = require('@google-cloud/logging');`
it always returns undefined.

Using regular injection on typescript with Angular and also tried in NestJS, only:
`import { Logging } from '@google-cloud/logging';`
returns an instance that I could work with but still the same exception gets thrown.

I'm gonna put together a quick repo and share it with you so you can check it out.

**Comment by arianlopezc** [comment]
@bcoe here is a repo that reproduces the problem I'm having:
https://github.com/arianlopezc/google-logging-test

**Comment by arianlopezc** [comment]
@bcoe that was one of the permutations that I tried. I tried it again but still, I'm facing the issue. Is there any other possible solution I could use? Maybe a different library or try to switch to make a call to the API using an Http request.

**Comment by arianlopezc** [comment]
hi @bcoe I tried the repo in another computer that I have and it works like a deer. For some reason in my regular machine it always throws me the exception. Do you get an idea on what could it be in my environment? For a moment I thought about the location of the file or something like that, so I moved it to different locations just to try out but still nothing.

**Comment by bcoe** [comment]
@arianlopezc off the top of my head, I would guess that the meta-information file being pointed to by `GOOGLE_APPLICATION_CREDENTIALS` is invalid, or corrupt.

I would try recreating a credentials file, ensuring it has access to logging, following the steps here:

https://cloud.google.com/docs/authentication/getting-started

Then, make sure you're explicitly running your application with the variable `GOOGLE_APPLICATION_CREDENTIALS` set, pointing to the absolute path of the credentials file, try dropping the `~/` prefix if you're using it.

**Comment by arianlopezc** [comment]
@bcoe I double checked all. The path to the file doesn't have any prefixes, I downloaded a new credentials JSON file, the environment variable is set, etc... still the same issue...
I just airdropped the JSON file to the other computer, set the path to the JSON, and all works fine...
For some reason in my personal machine there is something that is not right.
I don't know, I'm just gonna check that I don't have the same issues when I deploy it to my containers and if it is all good then I will close this issue because it makes no sense to keep worrying about something that feels like the issue is with my local environment.

**Comment by bcoe** [comment]
@arianlopezc I'm going to go ahead and close this, please let me know if you do figure out the discrepancy between your personal computer and other environments at some point (you've got me curious).

**Comment by jtyers** [comment]
@bcoe this occurs for me also. Like @arianlopezc I suspect a local environment issue.

Running the sample code you posted above results in the same error, even without any `GOOGLE_APPLICATION_CREDENTIALS` set (I am logged in via `gcloud` as the owner of the project I'm testing with, and those creds definitely work as I've been `terraform`ing other infrastructure with them).

If I set `GOOGLE_APPLICATION_CREDENTIALS` to a JSON credentials file for a service account with access to `roles/logging.logWriter`, I get the same error.

`(node:47026) UnhandledPromiseRejectionWarning: Error: Invalid response from metadata service: incorrect Metadata-Flavor header.`

**Comment by dustingraham** [comment]
I'm having this same error.
```
(node:25564) UnhandledPromiseRejectionWarning: Error: Unexpected error determining execution environment: Unexpected error determining execution environment: Invalid response from metadata service: incorrect Metadata-Flavor header.
```

I've checked environment variables, and configuration values. I suspect it's failing to pick up my credentials, but I've triple cross checked that it has the right file set in `keyFilename` but no luck. I'll keep trying, but it's frustrating that there isn't a better error message.

**Comment by arianlopezc** [comment]
I fully agree with you @dustingraham and with what @jtyers mentioned. As a way to validate that idea a bit, I cloned my repo on another machine and the whole thing worked just fine. There is some local configuration that either we are missing or have different that causes this issue.

**Comment by joshualyon** [comment]
Just wanted to add that I had this happen to me after factory resetting a PC and it turns out it was ATT's DNS catch-all that was causing issues (paired with the internal auth/metadata module's expectations around querying the metadata server). 

I compared each and every setting that I could think of between a machine where things were working and on one that wasn't. Everything seemed to match, so I stepped through the code line by line and realized that it was failing in `google-auth-library\build\src\auth\envDetect.js` where it tried to figure out what type of GCP environment, if any, the code is running in.

When it got to the `isComputeEngine()` check, it would call the `gcpMetadata.isAvailable()` method which attempts a network call to the metadata service and tries to fail fast if it's unavailable. Unfortunately, ATT's DNS 'Assist' that was catching all unknown DNS calls and returning a search result instead of just failing the network request. 

I changed my DNS entries on that PC over to Google's DNS (`8.8.8.8` and `8.8.4.4`) and the issue is now resolved for me.
